### PR TITLE
fix: Panic in UNION ALL queries

### LINKED ISSUE #4795: Panic on UNION ALL queries
### What type of bug is this?

Crash

### What subsystems are affected?

Query Engine

### Minimal reproduce step

```sql
mysql> SELECT 123 as a, 'h'  as b
    -> UNION ALL
    -> SELECT 456 as a, 'e'  as b
    -> UNION ALL
    -> SELECT 789 as a, 'l'  as b;
```

### What did you expect to see?

```
mysql> SELECT 123 as a, 'h'  as b
    -> UNION ALL
    -> SELECT 456 as a, 'e'  as b
    -> UNION ALL
    -> SELECT 789 as a, 'l'  as b;
+-----+---+
| a   | b |
+-----+---+
| 123 | h |
| 456 | e |
| 789 | l |
+-----+---+
3 rows in set (0.00 sec)
```

### What did you see instead?

```
mysql> SELECT 123 as a, 'h'  as b
    -> UNION ALL
    -> SELECT 456 as a, 'e'  as b
    -> UNION ALL
    -> SELECT 789 as a, 'l'  as b;
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: *** NONE ***
```

### What operating system did you use?

NA

### What version of GreptimeDB did you use?

0.9.3

### Relevant log output and stack trace

```bash
2024-10-05T03:18:36.630412Z ERROR on_query{protocol="mysql"}: common_telemetry::panic_hook: panicked at src/common/recordbatch/src/adapter.rs:363:9:
attempt to add with overflow backtrace=   0: common_telemetry::panic_hook::set_panic_hook::{{closure}}
             at /home/lei/workspace/greptimedb/src/common/telemetry/src/panic_hook.rs:37:25
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/alloc/src/boxed.rs:2077:9
      std::panicking::rust_panic_with_hook
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:799:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:656:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/sys_common/backtrace.rs:171:18
   4: rust_begin_unwind
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:652:5
   5: core::panicking::panic_fmt
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/panicking.rs:72:14
   6: core::panicking::panic_const::panic_const_add_overflow
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/panicking.rs:179:21
   7: <common_recordbatch::adapter::MetricCollector as datafusion_physical_plan::visitor::ExecutionPlanVisitor>::pre_visit
             at /home/lei/workspace/greptimedb/src/common/recordbatch/src/adapter.rs:363:9
   8: datafusion_physical_plan::visitor::visit_execution_plan
             at /home/lei/.cargo/git/checkouts/arrow-datafusion-8f64698f94ffb3e6/7823ef2/datafusion/physical-plan/src/visitor.rs:88:5
   9: datafusion_physical_plan::visitor::accept
             at /home/lei/.cargo/git/checkouts/arrow-datafusion-8f64698f94ffb3e6/7823ef2/datafusion/physical-plan/src/visitor.rs:30:9
  10: <common_recordbatch::adapter::RecordBatchStreamAdapter as futures_core::stream::Stream>::poll_next
             at /home/lei/workspace/greptimedb/src/common/recordbatch/src/adapter.rs:300:21
  11: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.30/src/stream.rs:120:9
  12: <query::metrics::OnDone<F> as futures_core::stream::Stream>::poll_next
             at /home/lei/workspace/greptimedb/src/query/src/metrics.rs:100:22
  13: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.30/src/stream.rs:120:9
  14: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/stream/stream/mod.rs:1638:9
  15: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/stream/stream/next.rs:32:9
  16: servers::mysql::writer::MysqlResultWriter<W>::write_query_result::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/writer.rs:164:75
  17: servers::mysql::writer::MysqlResultWriter<W>::try_write_one::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/writer.rs:113:93
  18: servers::mysql::writer::write_output::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/writer.rs:46:54
  19: <servers::mysql::handler::MysqlInstanceShim as opensrv_mysql::AsyncMysqlShim<W>>::on_query::{{closure}}::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/handler.rs:554:58
  20: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.40/src/instrument.rs:321:9
  21: <servers::mysql::handler::MysqlInstanceShim as opensrv_mysql::AsyncMysqlShim<W>>::on_query::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/handler.rs:421:5
  22: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/future/future.rs:123:9
  23: opensrv_mysql::AsyncMysqlIntermediary<B,R,W>::run::{{closure}}
             at /home/lei/.cargo/git/checkouts/opensrv-2d23bfb068524349/6bbc3b6/mysql/src/lib.rs:626:38
  24: opensrv_mysql::tls::plain_run_with_options::{{closure}}
             at /home/lei/.cargo/git/checkouts/opensrv-2d23bfb068524349/6bbc3b6/mysql/src/tls.rs:57:14
  25: servers::mysql::server::MysqlServer::do_handle::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/server.rs:206:68
  26: servers::mysql::server::MysqlServer::handle::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/server.rs:160:74
  27: servers::mysql::server::MysqlServer::accept::{{closure}}::{{closure}}::{{closure}}
             at /home/lei/workspace/greptimedb/src/servers/src/mysql/server.rs:142:82
  28: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/future/future.rs:123:9
  29: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:331:17
  30: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/loom/std/unsafe_cell.rs:16:9
      tokio::runtime::task::core::Core<T,S>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:320:13
  31: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:500:19
  32: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/panic/unwind_safe.rs:272:9
  33: std::panicking::try::do_call
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:559:40
  34: __rust_try
  35: std::panicking::try
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:523:19
  36: std::panic::catch_unwind
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panic.rs:149:14
  37: tokio::runtime::task::harness::poll_future
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:488:18
  38: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:209:27
  39: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:154:15
  40: tokio::runtime::task::raw::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:271:5
  41: tokio::runtime::task::raw::RawTask::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:201:18
  42: tokio::runtime::task::LocalNotified<S>::run
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/mod.rs:436:9
  43: tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:598:13
  44: tokio::runtime::coop::with_budget
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/coop.rs:107:5
      tokio::runtime::coop::budget
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/coop.rs:73:5
      tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:597:9
  45: tokio::runtime::scheduler::multi_thread::worker::Context::run
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:560:24
  46: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:513:21
  47: tokio::runtime::context::scoped::Scoped<T>::set
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context/scoped.rs:40:9
  48: tokio::runtime::context::set_scheduler::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context.rs:180:26
  49: std::thread::local::LocalKey<T>::try_with
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/thread/local.rs:283:12
  50: std::thread::local::LocalKey<T>::with
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/thread/local.rs:260:9
  51: tokio::runtime::context::set_scheduler
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context.rs:180:9
  52: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:508:9
  53: tokio::runtime::context::runtime::enter_runtime
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context/runtime.rs:65:16
  54: tokio::runtime::scheduler::multi_thread::worker::run
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:500:5
  55: tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:466:45
  56: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/task.rs:42:21
  57: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:331:17
  58: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/loom/std/unsafe_cell.rs:16:9
      tokio::runtime::task::core::Core<T,S>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:320:13
  59: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:500:19
  60: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/panic/unwind_safe.rs:272:9
  61: std::panicking::try::do_call
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:559:40
  62: __rust_try
  63: std::panicking::try
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:523:19
  64: std::panic::catch_unwind
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panic.rs:149:14
  65: tokio::runtime::task::harness::poll_future
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:488:18
  66: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:209:27
  67: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:154:15
  68: tokio::runtime::task::raw::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:271:5
  69: tokio::runtime::task::raw::RawTask::poll
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:201:18
  70: tokio::runtime::task::UnownedTask<S>::run
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/mod.rs:473:9
  71: tokio::runtime::blocking::pool::Task::run
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/pool.rs:160:9
  72: tokio::runtime::blocking::pool::Inner::run
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/pool.rs:518:17
  73: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/lei/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/pool.rs:476:13
  74: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/sys_common/backtrace.rs:155:18
  75: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/thread/mod.rs:542:17
  76: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/panic/unwind_safe.rs:272:9
  77: std::panicking::try::do_call
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:559:40
  78: __rust_try
  79: std::panicking::try
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panicking.rs:523:19
  80: std::panic::catch_unwind
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/panic.rs:149:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/thread/mod.rs:541:30
  81: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/core/src/ops/function.rs:250:5
  82: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/alloc/src/boxed.rs:2063:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/alloc/src/boxed.rs:2063:9
      std::sys::pal::unix::thread::Thread::new::thread_start
             at /rustc/72fdf913c53dd0e75313ba83e4aa80df3f6e2871/library/std/src/sys/pal/unix/thread.rs:108:17
  83: <unknown>
  84: <unknown>
```
```

### PR DESCRIPTION
**Pull request by v0y4g3r**
Improve MetricCollector by incrementing level and fix underflow issue; add tests for UNION ALL queries

I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- fixes #4795 

## What's changed and what's your intention?
Fix the panic in multi UNION ALL queries.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/common/recordbatch/src/adapter.rs [hunk 1]
```diff
@@ -365,8 +366,7 @@ impl ExecutionPlanVisitor for MetricCollector {
     }
 
     fn post_visit(&mut self, _plan: &dyn ExecutionPlan) -> std::result::Result<bool, Self::Error> {
-        // the last minus will underflow
-        self.current_level = self.current_level.wrapping_sub(1);
```
Comment (killme2008) [inline src/common/recordbatch/src/adapter.rs:369]: LGTM, but why change this line?

**Review by killme2008 — APPROVED** [review]

File: src/common/recordbatch/src/adapter.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/common/recordbatch/src/adapter.rs:369]: > LGTM, but why change this line?

The potention underflow here is caused by a missing `self.current_level += 1`. After this fix there should be no underflow anymore.

**Review by evenyag — APPROVED** [review]
Looks like it works.

@waynexia PTAL. I don't know if there are potential issues that I overlook.

**Review by waynexia — APPROVED** [review]
LGTM

**Pull request merged**